k8s_api_client/README.md

```diff
@@ -0,0 +1,162 @@
+# opa-kubernetes-api-client
+
+Rego functions to query the Kubernetes API directly from OPA.
+
+```rego
+package authz
+
+import data.kubernetes.api.client
+
+mutating_actions := {"update", "patch", "delete"}
+
+# Deny mutating action unless user is in group owning the resource
+deny[reason] {
+    mutating_actions[input.spec.resourceAttributes.verb]
+
+    cluster_resource := client.query_name_ns(
+        input.spec.resourceAttributes.resource,
+        input.spec.resourceAttributes.name,
+        input.spec.resourceAttributes.namespace,
+    )
+    resource_owner := cluster_resource.metadata.labels.owner
+    user_groups := {group | group := input.spec.user.group[_]}
+
+    not user_groups[resource_owner]
+
+    reason := sprintf("User %v not in group %v, has %v", [input.spec.user.name, resource_owner, user_groups])
+}
+```
+
+## Motivation
+
+Many policies interacting with Kubernetes depends on knowing the current state of the cluster. Examples of this include checking for hostname or paths already in use by ingress resources, the number of workloads running on a node, or the total amount of memory allocated to an application. Having this data available for decisions significantly increases the number of possible use cases for policy enforcement, whether it be for authorization or admission control.
+
+Systems that integrate OPA and Kubernetes commonly provide a [cached](https://github.com/open-policy-agent/kube-mgmt#caching) or [replicated](https://github.com/open-policy-agent/gatekeeper#replicating-data) view of the Kubernetes API in order to enable these type of policies. The eventually consistent views of the cluster are tolerable for most use cases, but not for all. This library provides helper functions to communicate directly with the kubernetes API from inside of Rego, allowing for a snapshot view of any resources deployed to the cluster.
```

Comment:
I'd probably change this paragraph to something like:

Systems that integrate OPA and Kubernetes commonly provide a [cached](https://github.com/open-policy-agent/kube-mgmt#caching) or [replicated](https://github.com/open-policy-agent/gatekeeper#replicating-data) view of the Kubernetes API in order to enable these type of policies. In some cases, replicating cluster state into OPA is overkill and it's more efficient to simply query for the required resource(s) from within the policy. This allows the policy to lookup the specific resource at policy evaluation-time. Note, because (by design) Kubernetes is eventually consistent, it's still possible for the API call to return stale results.

(Just to avoid giving people the impression that this provides a strongly consistent view of the cluster...)

Reply:
Good point. I've updated the paragraph, also adding that the two methods aren't mutually exclusive. I rephrased your suggestion slightly as I don't want it to be perceived that the only reason for using this is that caching might be overkill - the snapshot view offered by the kubernetes API vs that of a replicated view really meant the difference between usable and unusable kubernetes authorization webhooks in a real production systems where similar functions where used, so while the disclaimer is valid (and kept) I think it's still worth pointing out. I hope it's a fair compromise.